Flask application

iurygregory opened the pull request:
This PR add the Flask application that will provide the endpoint for Prometheus to access the metrics from all baremetal nodes

iurygregory committed:
Updates for Prometheus Exporter

- Split requirements in test-requirements and requirements
- Update .gitignore
- Initial Parser for IPMI metrics that was collected and tests
(This will be added in another PR: Cable, Drive, Processor, Battery,
 Module, Entity, Add-in, Critical, Voltage, OS, Event, Physical)
- Update Prometheus Driver to generate files with the metrics in
prometheus format
- Create simple driver to be used when we want to collect sample of
data

iurygregory committed:
Refactoring code

- Extract common functions between IPMI classes
- Update Prometheus driver to not use os.path.dirname in the files_dir
- Update driver test to use tearDown and remove the created files
- Remove unnecessary code in ipmi tests

iurygregory committed:
Updates for Driver, Parsers and Tests

-Added logging
-Changed `files_dir` option to `location` for Notification drivers
-Changed write to be done with block
-Extract all classes in ipmi to be functions
-Change manager to extract information based on the event type
-Collect ipmi information only for categories in the payload that
have a match function.
-Update tests for driver to use fixtures TempDir
-Update ipmi parser tests (all categories in one class, test the metric
name for each category)

iurygregory committed:
Update requirements and documentation

- Allow oslo.messaging version to be equal or higher than 9.4.0
- Update documentaion about how to verify if the driver is installed
and how to configure.

iurygregory committed:
Flask Application to Prometheus

- Update requirements and README
- Adds a Flask application that will merge the metrics file and
provide the result in the `/metrics` endpoint so Prometheus can
use it.

dtantsur commented on README.md:
````diff
@@ -14,23 +14,49 @@ $ python setup.py install
 3- Verify if the driver is installed
 ```
-$ python verify_installation.py
+$ pip install entry_point_inspector
````
Could you come up with a procedure that does not involve `pip install`? Or at least could you add `--user` to it to avoid recommending `sudo pip install` (although it won't help CentOS/RHEL users who don't have `pip`).

iurygregory replied:
Doug pointed out this was a good tool to use to verify the installation that`s why  I added.

iurygregory replied:
Added the --user

dtantsur commented on README.md:
````diff
@@ -14,23 +14,49 @@ $ python setup.py install
+```
+3- run the Flask Application
+```
+$ python -m flask run &
````
Please do not recommend this for production, people should use a real WSGI server.

iurygregory replied:
Ack I will see how to run in WSGI and update here

iurygregory replied:
I've added a NOTE "**NOTE:** if you want to deploy in production please check the Flask [documentation](http://flask.pocoo.org/docs/dev/deploying/)"

dtantsur replied:
I don't see a reason why we would recommend a non-production way at all...

dtantsur commented on ironic_prometheus_exporter/app/exporter.py:
```diff
@@ -0,0 +1,30 @@
+        DIR = config['oslo_messaging_notifications']['location']
+    except Exception as e:
+        LOG.error(e)
+        abort(404)
```
It's not necessary 404, any error will end up here. Although 404 is never correct for this endpoint, since there is no way users can provide something that cannot be found. It has to be 500, I guess.

iurygregory replied:
Ok o/

dtantsur commented on ironic_prometheus_exporter/app/exporter.py:
```diff
@@ -0,0 +1,30 @@
+        config.read(os.environ.get('IRONIC_CONFIG'))
+        DIR = config['oslo_messaging_notifications']['location']
+    except Exception as e:
+        LOG.error(e)
```
Logging requires a string as a first argument. I'd suggest:
```python
LOG.exception('Unexpected error')
```
(I'm not including `e` since `LOG.exception` outputs the complete traceback).

iurygregory replied:
Ack

dtantsur commented on ironic_prometheus_exporter/app/exporter.py:
```diff
@@ -0,0 +1,30 @@
+    content = ''
+    for file_name in all_files:
+        with open(file_name, 'r') as file:
+            content += ''.join(file.readlines())
```
Can we stream content as we generate it? Given that it's a string, it should not be hard.

iurygregory replied:
Going to look how to do this, but would be necessary? I think when Prometheus request he wan'ts to get all the content at once...

dtantsur replied:
Just to avoid building a potentially large string in memory.

dhellmann replied:
You could also use file.read() to skip splitting it into lines, since you're just joining them back together again anyway.

iurygregory replied:
I've switched to use the streaming from Flask  http://flask.pocoo.org/docs/1.0/patterns/streaming/

dtantsur commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,169 @@
+                else:
+                    values[entry] = 1
+        except Exception as e:
+            LOG.error(e)
```
This is a recipe for a debugging nightmare. At the very least you need to use `LOG.exception` so that the logs actually contain some context. Keep in mind that e.g. the message of a `KeyError` is just a key.

dtantsur replied:
The same applies to other `extract_labels`.

iurygregory replied:
I will change and also add more logs information to help with the debug

dtantsur commented on ironic_prometheus_exporter/app/exporter.py:
```diff
@@ -0,0 +1,30 @@
+import logging
+import os
+
+from flask import abort, Flask
```
Do we really need `Flask` to expose one endpoint without any routing and any input processing? I guess you could use `werkzeug` or anything similar directly. E.g. for our JSON RPC implementation I used very low level `WebOp` library: https://github.com/openstack/ironic/blob/master/ironic/common/json_rpc/server.py#L102.

iurygregory replied:
The only thing that will be doing requests for the Flask Application will be Prometheus, so I think this is fine

dtantsur replied:
What I'm trying to say is that Flask is a vast overkill here.

dhellmann replied:
Flask probably is overkill, but it's not a huge resource consumer and this is working so it seems fine to me. If we were going to rewrite it, I would suggest doing that in golang to make distribution simpler and to be consistent with other kubernetes components.

dtantsur commented on ironic_prometheus_exporter/app/exporter.py:
```diff
@@ -0,0 +1,30 @@
+                 if os.path.isfile(os.path.join(DIR, name))]
+    content = ''
+    for file_name in all_files:
+        with open(file_name, 'r') as file:
```
Probably not important on our scale.. but I'm not sure writing to files in python is atomic. You may end up reading half-written files. We can worry about it later, but let's keep it in mind.

iurygregory replied:
I will add a note so we can remember about this

dhellmann replied:
When the files are created we should write to files with different names and then rename the file (that operation is atomic).

iurygregory replied:
The oslo driver writes a file per node (`node_name`), so we should rename the files with some kind of prefix after written and Flask will only read files with the prefix?
like: `node1` we rename to `metrics_node1`, flask only read files in the directory that starts with `metrics_` ?

dhellmann replied:
Maybe write to a temporary directory and then move the files into the real directory?

iurygregory replied:
Ok, I'll update the driver to do this, since he is that write the files

dtantsur commented on ironic_prometheus_exporter/tests/test_driver.py:
```diff
@@ -8,12 +12,66 @@ class TestPrometheusFileNotifier(test_utils.BaseTestCase):
+                     if os.path.isfile(os.path.join(DIR, name))]
+        self.assertEqual(node1, node2)
+        self.assertEqual(len(all_files), 1)
+        self.assertTrue(node1 in all_files)
```
Please use `assertIn` (the same applies below).

iurygregory replied:
Ack

dtantsur commented on ironic_prometheus_exporter/tests/test_ipmi_parser.py:
```diff
@@ -0,0 +1,130 @@
+        management_metrics_name = ipmi.metric_names(self.payload['Management'],
+                                                    'baremetal_', '')
+        self.assertEqual(len(management_metrics_name), 1)
+        self.assertTrue('baremetal_front_led_panel' in management_metrics_name)
```
ditto re `assertIn`

iurygregory replied:
Ack

iurygregory committed:
Review Updates

- Updated documentation:
 * install entry_point_inspector with --user
 * added NOTE to check how to deploy the application in production

- Flask application:
 * abort with 500 code
 * stream the files to avoid a large string in memory

- Prometheus Notifier
 * write node information to temporary file and after finish rename to
   the correct file since rename is an atomic operation.

- IPMI
 * added LOG.info for functions to help when debugging
 * changed LOG.error to LOG.exception
 * refactor metric_names to use kwargs for option arguments
 * refactor all category functions to a common one that uses
   a global dictionary with the necessary parameters to call
   the right function for each scenario.

- Manager
 * Update to use the common function from IPMI

- Tests
 * Update tests to match the modifications in the functions

simonpasquier commented:
FWIW, the client_python library can integrate directly with [Flask](https://github.com/prometheus/client_python/#flask). An advantage is that this library is one of the reference implementations for [OpenMetrics](https://openmetrics.io/) (the standardized version of the Prometheus exposition format) so leveraging it means that we are less likely to encounter interoperability issues. Another option also mentioned would be to write a Go application: it should be straightforward (probably a hundred lines of code) and also a good guarantee for interop.

iurygregory committed:
IPMI changes

- added HELP header for the metrics (for now we will have no
description)
- fixed typos
- added Entity ID as label
- added regex to ensure metric_name is valid
- Sensor ID will be used the way it is.

iurygregory committed:
Move to client_python from Prometheus

- update oslo.notifier and ipmi parser to use the prometheus client_python
- update tests to use prometheus client_python
- removal of the manager since doesn't seems to be necessary
- update flask application to run under uWSGI using command line
- update requirements to use the prometheus client_python and uwsgi
- update README with information about how to run under uWSGI

simonpasquier commented on ironic_prometheus_exporter/app/exporter.py:
```diff
@@ -0,0 +1,31 @@
+
+
+@application.route('/metrics', methods=['GET'])
+def prometheus_metrics():
```
Though this is probably for future work, it could be interesting to leverage [`make_wsgi_app`](https://github.com/prometheus/client_python/blob/master/prometheus_client/exposition.py#L36) as it will do the right thing based on the Prometheus request (eg choosing the right encoder, legacy or [OpenMetrics](https://openmetrics.io/)).

In the same vein, using a custom collector would allow to inject [process](https://github.com/prometheus/client_python#process-collector), [Python](https://github.com/prometheus/client_python#platform-collector) and GC metrics. It would also make sure that the read data is valid (eg avoid serving corrupted files). See [`text_string_to_metric_families`](https://github.com/prometheus/client_python/blob/2206f15991e0a0cb6454b71fbe93f1394e0ae90f/prometheus_client/parser.py#L17) and [Custom collector](https://github.com/prometheus/client_python/blob/2206f15991e0a0cb6454b71fbe93f1394e0ae90f/README.md#custom-collectors) for details.

Without testing, it would look like this:

```python
from prometheus_client.parser import text_string_to_metric_families
from prometheus_client.exposition import choose_encoder
...

class IPMICollector(object):
    def __init__(self, dir):
       self.dir = dir

    def collect(self):
        # Expose  the modification time of every metrics file.
        mtime = GaugeMetricFamily('ironic_exporter_file_mtime_seconds', 'Unixtime mtime of file successfully read.', labels=["file"])
        all_files = [os.path.join(self.dir, name) for name in os.listdir(self.dir)
                 if os.path.isfile(os.path.join(self.dir, name))]
        
        for file_name in all_files:
            with open(file_name, 'r') as file:
                text_string_to_metric_families(file.read())
                stats = os.stat(filename)
                yield mtime.labels(file=file_name).set(stats.st_mtime)

# Register the custom collector with the default registry which already collects process, platform and GC metrics.
REGISTER.register(IPMICollector(DIR))

@application.route('/metrics', methods=['GET'])
def prometheus_metrics():
    encoder, content_type = choose_encoder(request.headers.get('HTTP_ACCEPT'))
    return Response(encoder(), headers=Headers([('Content-Type', content_type)]))
```

dtantsur replied:
++ (can be done in a follow-up, I think)

simonpasquier reviewed (commented):
LGTM for the usage of prometheus_client. Added some ideas for improvements that can probably be addressed later.

dtantsur commented on README.md:
````diff
@@ -14,23 +14,58 @@ $ python setup.py install
+```
+
+**Running under uWSGI**
+Reproduce the Steps 1 and 2 (You don't need to set `FLASK_APP` variable) and run the command below:
````
Let us leave only this one (probably add `yum install uwsgi-python`).

dtantsur commented on ironic_prometheus_exporter/app/exporter.py:
```diff
@@ -0,0 +1,31 @@
+    try:
+        config = configparser.ConfigParser()
+        config.read(os.environ.get('IRONIC_CONFIG'))
+        DIR = config['oslo_messaging_notifications']['location']
```
nit: s/DIR/location/

dtantsur commented on ironic_prometheus_exporter/app/exporter.py:
```diff
@@ -0,0 +1,31 @@
+        config.read(os.environ.get('IRONIC_CONFIG'))
+        DIR = config['oslo_messaging_notifications']['location']
+    except Exception:
+        LOG.error('Unexpected error')
```
s/error/exception/ (to log traceback)

dtantsur commented on ironic_prometheus_exporter/app/exporter.py:
```diff
@@ -0,0 +1,31 @@
+def prometheus_metrics():
+    try:
+        config = configparser.ConfigParser()
+        config.read(os.environ.get('IRONIC_CONFIG'))
```
I wonder if we need to cache the result of this to avoid reading the config file on each request.

dtantsur commented on ironic_prometheus_exporter/app/exporter.py:
```diff
@@ -0,0 +1,31 @@
+        abort(500)
+
+    all_files = [os.path.join(DIR, name) for name in os.listdir(DIR)
+                 if os.path.isfile(os.path.join(DIR, name))]
```
nit: you don't need a list, you can use a generator `(...)` instead

dtantsur commented on ironic_prometheus_exporter/app/wsgi.py:
```diff
@@ -0,0 +1,4 @@
+from ironic_prometheus_exporter.app.exporter import application
+
+if __name__ == '__main__':
+    application.run()
```
This is present twice: in exporter.py and here

dtantsur commented on ironic_prometheus_exporter/messaging.py:
```diff
@@ -18,13 +23,35 @@ class PrometheusFileDriver(notifier.Driver):
+                nodeFile = os.path.join(self.location, node_name)
+                write_to_textfile(nodeFile, registry)
+        except Exception as e:
+            LOG.error(e)
```
s/error/exception/ to log traceback

dtantsur commented on ironic_prometheus_exporter/messaging.py:
```diff
@@ -18,13 +23,35 @@ class PrometheusFileDriver(notifier.Driver):
-        with open(self.file_path, 'w') as prometheus_file:
-            json.dump(message, prometheus_file)
+        try:
+            if message['event_type'] == 'hardware.ipmi.metrics':
```
nit:
```python
if message['event_type'] != 'hardware.ipmi.metrics':
    return
```

dtantsur commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,166 @@
+def metric_names(payload, prefix, sufix, **kwargs):
+    LOG.info('metric_names function called with payload=%s' % str(payload))
+    LOG.info('prefix=%s | sufix=%s | kwargs=%s' %
+             (prefix, sufix, str(kwargs.items())))
```
These two things should be debug, not info

dtantsur commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,166 @@
+
+    returns: a dictionarty of dictionaries {<entry>: {label_name: label_value}}
+    """
+    LOG.info('extract_labels function called with: entries=%s | payload=%s | \
```
ditto re logging (probably more instances below)

dtantsur commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,166 @@
+                            'sensor_id': sensor_id}
+            entries_labels[entry] = metric_label
+        except Exception as e:
+            LOG.exception(e)
```
For `LOG.exception` you need to provide context, the exception itself will be logged anyway. So something like
```python
LOG.exception("Failed to process entry %s", entry)
```
The same applies to where I commented about s/error/exception/ above

dtantsur commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,166 @@
+                else:
+                    values[entry] = 1
+        except Exception as e:
+            LOG.exception(e)
```
ditto re logging

dtantsur commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,166 @@
+                sensor_values = payload[entry]['Sensor Reading'].split()
+                if not use_ipmi_format:
+                    if not re.search(r'(\d+(\.\d*)?|\.\d+)', sensor_values[0]):
+                        raise Exception("No valid value in Sensor Reading")
```
```python
raise Exception("Invalid value in sensor reading entry %s: %s", entry, sensor_values[0])
```

dtantsur commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,166 @@
+                                use_ipmi_format=use_ipmi_format)
+        if all(v is None for v in values.values()):
+            continue
+        g = Gauge(metric, '', labelnames=labels.get(entries[0]).keys(),
```
nit: `keys()` is an iterator in python 3 and is usually redundant with dicts. so probably something like `labelnames=list(labels.get(entries[0]))`

dtantsur commented on ironic_prometheus_exporter/tests/test_driver.py:
```diff
@@ -8,12 +12,66 @@ class TestPrometheusFileNotifier(test_utils.BaseTestCase):
+        all_files = [name for name in os.listdir(DIR)
+                     if os.path.isfile(os.path.join(DIR, name))]
+        self.assertEqual(node1, node2)
+        self.assertEqual(len(all_files), 1)
```
nit: maybe
```python
self.assertEqual({node1, node2}, set(all_files))
```
?

dtantsur commented on requirements.txt:
```diff
@@ -1,5 +1,6 @@
-oslo.messaging!=9.0.0 # Apache-2.0
-stestr>=2.0.0 # Apache-2.0
+oslo.messaging>=9.4.0 # Apache-2.0
+uWSGI # Apache-2.0
```
This is not a requirement of the project, it should not be here.

dtantsur reviewed (approved):
Quite a few comments inline, but I think everything can be fixed in a follow-up. Thanks!

iurygregory commented:
Thanks everyone, I will merge this so we can work on the ironic container and I will push other PR to update the reviews.
